vllm/worker/multi_step_model_runner.py

```diff
@@ -0,0 +1,493 @@
+            assert seq_group.seq_len is None  # Decode
+            assert seq_group.query_len is None  # Decode
+
+    def _advance_step(
+            self, model_input: MutableModelInputForGPUWithMultiStepMetadata,
+            out: SamplerOutput
+    ) -> MutableModelInputForGPUWithMultiStepMetadata:
+        frozen_model_input = model_input.frozen_model_input
+        assert frozen_model_input is not None
+        assert frozen_model_input.attn_metadata is not None
+
+        num_seqs = model_input.num_seqs
+        num_queries = model_input.num_queries
+        assert num_seqs > 0
+        assert num_queries > 0
+        assert num_seqs >= num_queries
+
+        attn_metadata = frozen_model_input.attn_metadata
+        assert isinstance(attn_metadata, FlashAttentionMetadata)
+        self._update_flash_attn_metadata(attn_metadata, num_seqs, num_queries)
+
+        # Update GPU tensors
+        ops.advance_step(
+            num_seqs=num_seqs,
+            num_queries=num_queries,
+            block_size=self.block_size,
+            input_tokens=frozen_model_input.input_tokens,
+            sampled_token_ids=model_input.outputs[-1].sampled_token_ids,
+            input_positions=frozen_model_input.input_positions,
+            seq_lens=attn_metadata.seq_lens_tensor,
+            slot_mapping=attn_metadata.slot_mapping,
+            block_tables=attn_metadata.block_tables)
+
+        if frozen_model_input.seq_lens is not None:
+            for i in range(num_queries):
+                frozen_model_input.seq_lens[i] = attn_metadata.seq_lens[i]
+
+        return model_input
+
+
+def _pythonize_sampler_output(
+        model_input: MutableModelInputForGPUWithMultiStepMetadata,
+        output: SamplerOutput, pinned_sampled_token_buffer: torch.Tensor,
+        sampled_token_ids: Optional[torch.Tensor]) -> SamplerOutput:
+    """ This function is only called when the output tensors are ready. 
+    See ModelOutput
+    """
+
+    assert sampled_token_ids is not None
+    assert model_input.frozen_model_input is not None
+
+    frozen_model_input = model_input.frozen_model_input
+    assert frozen_model_input.sampling_metadata is not None
+    # samples generation should have been skipped
+    assert not output.outputs
+
+    pinned_buffer = pinned_sampled_token_buffer[:model_input.num_queries]
+
+    # CPU GPU sync
+    pinned_buffer = pinned_buffer.copy_(sampled_token_ids, non_blocking=False)
+
+    # this will not block as the tensors are already on CPU
+    samples_list = pinned_buffer.tolist()
+
+    sampling_metadata = frozen_model_input.sampling_metadata
+
+    for (seq_group, sample_result) in zip(sampling_metadata.seq_groups,
+                                          samples_list):
+        seq_ids = seq_group.seq_ids
+        next_token_ids = sample_result
+        parent_ids = [0]
+        seq_outputs: List[SequenceOutput] = []
+        assert len(seq_group.sampling_params.logits_processors) == 0, (
```

Comment:
Nice, thank you. Maybe add to a backlog or tracking issue somewhere to support this in the future

Reply:
Have one ongoing at the bottom of the PR

> Follow up work:

> add ogprob support in _pythonize_sampler_output
> support chunked-prefill
> support guided decoding
> add flag to enforce synchronous pythonization (for logit processors and guided decoding)
> support spec-decode
> support prefix caching